pkg/controller/connectionsecret/connectionsecrets.go

```diff
@@ -20,6 +22,35 @@ import (
 
 const ConnectionSecretsEnsuredEvent = "ConnectionSecretsEnsured"
 
+func ReapOrphanConnectionSecrets(ctx context.Context, k8sClient client.Client, projectID, namespace string, projectDeploymentNames []string) ([]string, error) {
+	secretList := &corev1.SecretList{}
+	labelSelector := labels.SelectorFromSet(labels.Set{TypeLabelKey: CredLabelVal, ProjectLabelKey: projectID})
+	err := k8sClient.List(context.Background(), secretList, &client.ListOptions{
+		LabelSelector: labelSelector,
+		Namespace:     namespace,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed listing possible orphan secrets: %w", err)
+	}
+
+	removedOrphanSecrets := []string{}
+	for _, secret := range secretList.Items {
+		clusterName, ok := secret.Labels[ClusterLabelKey]
+		if !ok {
+			continue
+		}
+		if clusterExists := stringutil.Contains(projectDeploymentNames, clusterName); clusterExists {
+			continue
+		}
+		if err := k8sClient.Delete(ctx, &secret); err != nil {
```

Comment:
the method `ReapOrphanConnectionSecrets` is doing more than the reader is assuming. it is also causing the side-effect of deleting secrets. Either we:
1. factor deletion in its own function, not called in a `Read...` function
2. Rename this function to something that informs the reader that a deletion happens here as well